### PR TITLE
Bug 577923 - Allow hiding scope items in marker support filter dialog

### PR DESCRIPTION
**Pull request by eobrienPilz**
Moving patch from Gerrit: https://git.eclipse.org/r/c/platform/eclipse.platform.ui/+/191618

Bug 577923 - Irrelevant items cannot be removed/hidden from scope area
in the filter used by MarkerSupportView.

- Added the possibility to make radio buttons hidden using the
markerContentGenerator.

- Added 2 tests that items can be hidden and a second that default
behavour is to show all items.

- Rebased branch to see if it helps fix build fail.

Change-Id: Ied05d59195d58bc96b2e41ee5499a4fa164da837
Signed-off-by: Enda O Brien <E.OBrien@pilz.ie>

### CODE REVIEW

**Review by wimjongman — CHANGES_REQUESTED** [review]
This patch looks great. Thanks, @eobrienPilz 

I am only worried about the failing build. Maybe it is a setup issue. Can you push a little change to restart the build?

**Comment by akurtakov** [comment]
Probably rebase on master is what is needed.

**Comment by eobrienPilz** [comment]
Is there a recommended way to perform the rebase? I can rebase locally to master but then I think I may need to use the --force flag when pushing.

**Comment by iloveeclipse** [comment]
The way I'm working is: 
- checkout "master" branch and pull from origin repo to master
- create branch on top of master "issue_42" (set "rebase" policy in the new branch dialog in Eclipse)
- hack & commit on that "issue_42" branch
- push  "issue_42" branch to "your" repo & create PR from that
- hack & amend commit on that branch if I need any changes
- checkout "master" and pull from origin to get latest commits
- checkout "issue_42", in history view right click "master" and say "Rebase head on"
- **force** push  "issue_42" branch to "your" repo (PR will be updated automatically)

**Comment by eobrienPilz** [comment]
Would it work if I:
- checkout master and pull
- checkout my branch
- rebase my branch with "git rebase master"
- force push to my branch "git push origin myBranch --force"

**Comment by eobrienPilz** [comment]
I tried the rebase and force push but build still fails. 

@iloveeclipse  What do you mean when you refer to issue_42 branch ?

**Comment by eobrienPilz** [comment]
It seems to be building ok now after another rebase. Is it ok to land ?

File: bundles/org.eclipse.ui.ide/src/org/eclipse/ui/views/markers/internal/ContentGeneratorDescriptor.java
Comment (iloveeclipse) [inline bundles/org.eclipse.ui.ide/src/org/eclipse/ui/views/markers/internal/ContentGeneratorDescriptor.java]: If I see it right, this would silently peek some random contribution, in case multiple exists. Surely nothing we want. 

This "extension" looks more like a product customization and not like a typical extension point. Have you considered to use preferences service & product customization instead of introducing extension point that can only be extended by a single bundle?

Comment (eobrienPilz) [reply bundles/org.eclipse.ui.ide/src/org/eclipse/ui/views/markers/internal/ContentGeneratorDescriptor.java]: There shouldn't be problems with multiple occurrences.
a specific markerContentGenerator is given in the constructor of MarkerSupportView.
markerScopeAreaConfiguration that I added is optionally part of markerContentGenerator (0 or 1 occurrences).
No change in behavior for existing clients if markerScopeAreaConfiguration not present.

I haven't looked at preferences service & product customization but this extension point was already there. 
I just added a small piece for configuring the scope area.

Comment (eobrienPilz) [reply bundles/org.eclipse.ui.ide/src/org/eclipse/ui/views/markers/internal/ContentGeneratorDescriptor.java]: Are there any other concerns or can this be merged now ?

Comment (eobrienPilz) [reply bundles/org.eclipse.ui.ide/src/org/eclipse/ui/views/markers/internal/ContentGeneratorDescriptor.java]: Just to add in relation to using the preference service. I don't think this is the right way to fix the issue.
If I understand correctly what you mean it would require adding a set of key / value pairs to plugin_customization.ini to remove certain options from the scope are of the filter dialog.

Reasons I prefer my solution.

- It wont be obvious to users of the MarkerSupportView that these options exist. In contrast they will already be providing an implementation of org.eclipse.ui.ide.markerSupport so this will be the first place to look for options to remove from scope area.
- No encapsulation with key/value pairs. For our use case we want to remove 3 options. Other use cases may differ.
- End users may toggle the flags in plugin_customization.ini exposing the confusing options.

**Comment by akurtakov** [comment]
What is the status of this one guys?

**Comment by eobrienPilz** [comment]
I'd like it to be merged unless there is something else that needs to be addressed ?

**Comment by jukzi** [comment]
@iloveeclipse @mickaelistria  please finish review

**Comment by jukzi** [comment]
@wimjongman  please finish your review

**Comment by wimjongman** [comment]
> @wimjongman please finish your review

Hey Jorg. Thanks for pinging me. It has been a long time since I looked at this, so I need time to rewrap my head around it. I am at a two-week conference at the moment. So there's no time in the immediate future. If someone else can bring this in, then that would be great. Otherwise, I will reply in about three weeks.

**Comment by eobrienPilz** [comment]
> > @wimjongman please finish your review
> 
> Hey Jorg. Thanks for pinging me. It has been a long time since I looked at this, so I need time to rewrap my head around it. I am at a two-week conference at the moment. So there's no time in the immediate future. If someone else can bring this in, then that would be great. Otherwise, I will reply in about three weeks.

@jukzi Is there somebody else you can assign to finish the review ?

**Comment by jukzi** [comment]
> @jukzi Is there somebody else you can assign to finish the review ?

Every commiter is free to work on whatever he likes. We can't force each other to show interesst in anything. You may create interest by joining IDE workinggroup, directly contracting a commiter or working towards getting commiter rights yourself by first working on topics others are interested in too.

**Comment by eobrienPilz** [comment]
> > @jukzi Is there somebody else you can assign to finish the review ?
> 
> Every commiter is free to work on whatever he likes. We can't force each other to show interesst in anything. You may create interest by joining IDE workinggroup, directly contracting a commiter or working towards getting commiter rights yourself by first working on topics others are interested in too.

Do you know where I can find more info on  getting committer rights ?

**Comment by jukzi** [comment]
> Do you know where I can find more info on getting committer rights ?

@vogella can you help?

**Comment by eobrienPilz** [comment]
@iloveeclipse @mickaelistria  : If I was to reimplement this using the preference service as was suggested earlier by @iloveeclipse would either of you be willing to review ?

**Comment by eobrienPilz** [comment]
I pushed a change to the pr.  The scope items can be hidden using a preference. The extension point changes are reverted.
@iloveeclipse @mickaelistria  Would either of you have a chance to review ?

**Review by wimjongman — COMMENTED** [review]
The code looks good but I do not see the preference page for these new settings. Is this supposed to be a hidden feature and these preferences are to be set programmatically?

If so, then where is the documentation for anyone else interested in this?

**Comment by wimjongman** [comment]
> > > @jukzi Is there somebody else you can assign to finish the review ?
> > 
> > 
> > Every commiter is free to work on whatever he likes. We can't force each other to show interesst in anything. You may create interest by joining IDE workinggroup, directly contracting a commiter or working towards getting commiter rights yourself by first working on topics others are interested in too.
> 
> Do you know where I can find more info on getting committer rights ?

You have to get noticed by providing a bunch of quality patches and showing commitment to the project. Eventually you will be  invited.

**Comment by eobrienPilz** [comment]
> The code looks good but I do not see the preference page for these new settings. Is this supposed to be a hidden feature and these preferences are to be set programmatically?
> 
> If so, then where is the documentation for anyone else interested in this?

Yes, it's a hidden feature allowing preferences to be set programmatically. I added a javadoc comment on the new keys in IDEInternalPreferences.java. Developers should be able to find them with a little debugging. It looks to be similar for some other keys in the same file that I have checked.

**Review by mickaelistria — COMMENTED** [review]
The code looks sane, but given the context:
Wouldn't it be simpler to add a `protected ScopeArea createScopeArea()` in `FiltersConfigurationDialog` since you're already overriding it? With it, you could easily hook in your implementation without introducing preferences in Platform.

I'm telling that because "internal preferences" are usually a strange pattern, and also because it looks like this the need for filtering you have is not a general purpose use-case. The risk would be that some people change preferences, and it also requires you to play with plugin_customization.ini in your product just to tweak a view of yours, while it's usually not the most convenient and portable approach (ie someone using your plugin/view without using the product wouldn't get the filters).
If a new internal protected method works for you (as it seems it did earlier), it may be easier for everyone to continue it. What do you think?

**Comment by BeckerWdf** [comment]
> I'm telling that because "internal preferences" are usually a strange pattern

I fully agree. I like Mickael's proposal.

**Comment by eobrienPilz** [comment]
Yes, the way I see it working with a new createScopeArea protected method is ScopeArea.createContents can be overridden to provide custom contents, i.e. less buttons for my case.
In the existing ScopeArea implementation, the buttons (and maybe workingSetArea) field needs to be made protected so that buttons can be created from the overridden implementation.
Changes from this mr to handle null buttons would also need to be kept.

**Comment by eobrienPilz** [comment]
> Yes, the way I see it working with a new createScopeArea protected method is ScopeArea.createContents can be overridden to provide custom contents, i.e. less buttons for my case. In the existing ScopeArea implementation, the buttons (and maybe workingSetArea) field needs to be made protected so that buttons can be created from the overridden implementation. Changes from this mr to handle null buttons would also need to be kept.

... actually, looking more closely a bit more is needed. ScopeArea is package private so would need to be made public if its to be overridden. 
**Alternatively** maybe a better solution GroupFilterConfigurationArea could be made public and a completely new ScopeArea implementation provided. This way the existing ScopeArea can stay as it is.
MarkerFieldFilterGroup will also need to be public to provide a GroupFilterConfigurationArea implementation.
Any opinions ?

**Comment by mickaelistria** [comment]
> Any opinions ?

Any of those proposals can probably be implemented. If there is one that is technically superior to the others, let's adopt it. Which one would you recommend?

**Comment by eobrienPilz** [comment]
Thinking a bit more, the new createScopeArea method in FiltersConfigurationDialog should probably return a ScopeArea rather than a GroupFilterConfigurationArea. If FiltersConfigurationDialog needs something specific from ScopeArea in future it can get it without breaking the API. 

My preference would be to make ScopeArea, MarkerFieldFilterGroup  + constants for button indexes public and allow the client to provide the custom scope area implementation. In the existing ScopeArea implementation buttons and scope fields should be protected. I think the null checks that were there for preferences solution aren't needed. How does it sound to you ?

**Comment by eobrienPilz** [comment]
Also, just to add that from my point of view it would be nice at some point in future if this code could be moved out of .internal packages, so that there is some reassurance it wont go away in future. Though, I am not proposing that as part of this mr.

**Comment by eobrienPilz** [comment]
> Thinking a bit more, the new createScopeArea method in FiltersConfigurationDialog should probably return a ScopeArea rather than a GroupFilterConfigurationArea. If FiltersConfigurationDialog needs something specific from ScopeArea in future it can get it without breaking the API.
> 
> My preference would be to make ScopeArea, MarkerFieldFilterGroup + constants for button indexes public and allow the client to provide the custom scope area implementation. In the existing ScopeArea implementation buttons and scope fields should be protected. I think the null checks that were there for preferences solution aren't needed. How does it sound to you ?

Should I push an implementation for this ?

**Comment by wimjongman** [comment]
> Should I push an implementation for this ?

Yes, go ahead.

**Comment by eobrienPilz** [comment]
I am not sure why some of the checks are failing. The branch point is green now but the errors in "CodeQL call / Test Results" don't appear to be related to my changes.

**Comment by eobrienPilz** [comment]
Anyone have a chance to review this  ?

File: bundles/org.eclipse.ui.ide/src/org/eclipse/ui/internal/views/markers/FiltersConfigurationDialog.java
Comment (wimjongman) [inline bundles/org.eclipse.ui.ide/src/org/eclipse/ui/internal/views/markers/FiltersConfigurationDialog.java:188]: Some javadoc would be nice

**Review by wimjongman — CHANGES_REQUESTED** [review]

File: bundles/org.eclipse.ui.ide/src/org/eclipse/ui/internal/views/markers/FiltersConfigurationDialog.java
Comment (eobrienPilz) [reply bundles/org.eclipse.ui.ide/src/org/eclipse/ui/internal/views/markers/FiltersConfigurationDialog.java:188]: Javadoc is added for that method.

**Comment by wimjongman** [comment]
Since there is now new API there should be a @Since tag for all new API? @iloveeclipse? Other that that this looks fine to me.

did you already implement it?

**Comment by eobrienPilz** [comment]
> Since there is now new API there should be a @SInCE tag for all new API? @iloveeclipse? Other that that this looks fine to me.
> 
> did you already implement it?

@iloveeclipse  I'd also be interested to know if @since tag is required. I thought that maybe since these changes are to code that is in internal packages maybe its not really API. But I can add it if required.

@wimjongman "> did you already implement it?" - not sure what this question refers to or if its for me ?

**Comment by iloveeclipse** [comment]
Briefly checked the code. The new test should be added to the testsuite, otherwise it is not run. All changes are on internal classes, so I assume no API is changed and no since tags needed.

**Comment by wimjongman** [comment]
> @wimjongman "> did you already implement it?" - not sure what this question refers to or if its for me ?

Did you implement the changes in your RCP and are you happy with the result? Please make sure that this is all you need before we push it.

**Comment by eobrienPilz** [comment]
> > @wimjongman "> did you already implement it?" - not sure what this question refers to or if its for me ?
> 
> Did you implement the changes in your RCP and are you happy with the result? Please make sure that this is all you need before we push it.

Ah yes, good point. The test is basically doing what we need from our RCP application. I also did a manual visual test with the SDK problem view.

**Comment by iloveeclipse** [comment]
Two smaller proposals: 
1) Could you please squash all your changes to a *single* commit?
2) Could you please rebase that single commit on latest master state?

**Comment by eobrienPilz** [comment]
That is rebased to master and squashed now. Still some failing checks but looking at the details I don't see anything that relates to my changes.

**Comment by iloveeclipse** [comment]
> Still some failing checks but looking at the details I don't see anything that relates to my changes.

Not sure, I've not seen yet LeakTest failing ann all 3 platforms: https://github.com/eclipse-platform/eclipse.platform.ui/runs/23754171935

```
Test leaked modal shell: [Filters] expected:<0> but was:<1>
junit.framework.AssertionFailedError: Test leaked modal shell: [Filters] expected:<0> but was:<1>
	at junit.framework.Assert.fail(Assert.java:57)
	at junit.framework.Assert.failNotEquals(Assert.java:329)
	at junit.framework.Assert.assertEquals(Assert.java:78)
	at junit.framework.Assert.assertEquals(Assert.java:234)
	at junit.framework.TestCase.assertEquals(TestCase.java:377)
	at org.eclipse.ui.tests.harness.util.UITestCase.tearDown(UITestCase.java:227)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
```

**Comment by eobrienPilz** [comment]
> > Still some failing checks but looking at the details I don't see anything that relates to my changes.
> 
> Not sure, I've not seen yet LeakTest failing ann all 3 platforms: https://github.com/eclipse-platform/eclipse.platform.ui/runs/23754171935
> 
> ```
> Test leaked modal shell: [Filters] expected:<0> but was:<1>
> junit.framework.AssertionFailedError: Test leaked modal shell: [Filters] expected:<0> but was:<1>
> 	at junit.framework.Assert.fail(Assert.java:57)
> 	at junit.framework.Assert.failNotEquals(Assert.java:329)
> 	at junit.framework.Assert.assertEquals(Assert.java:78)
> 	at junit.framework.Assert.assertEquals(Assert.java:234)
> 	at junit.framework.TestCase.assertEquals(TestCase.java:377)
> 	at org.eclipse.ui.tests.harness.util.UITestCase.tearDown(UITestCase.java:227)
> 	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
> ```

Yes, it does appear to have just started failing now. Strangely if I run it locally it passes. Also it seems from looking at the test that its unrelated to any of my changes. Could there be instability in the test that is causing sporadic fails ?

**Comment by eobrienPilz** [comment]
> > > Still some failing checks but looking at the details I don't see anything that relates to my changes.
> > 
> > 
> > Not sure, I've not seen yet LeakTest failing ann all 3 platforms: https://github.com/eclipse-platform/eclipse.platform.ui/runs/23754171935
> > ```
> > Test leaked modal shell: [Filters] expected:<0> but was:<1>
> > junit.framework.AssertionFailedError: Test leaked modal shell: [Filters] expected:<0> but was:<1>
> > 	at junit.framework.Assert.fail(Assert.java:57)
> > 	at junit.framework.Assert.failNotEquals(Assert.java:329)
> > 	at junit.framework.Assert.assertEquals(Assert.java:78)
> > 	at junit.framework.Assert.assertEquals(Assert.java:234)
> > 	at junit.framework.TestCase.assertEquals(TestCase.java:377)
> > 	at org.eclipse.ui.tests.harness.util.UITestCase.tearDown(UITestCase.java:227)
> > 	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
> > ```
> 
> Yes, it does appear to have just started failing now. Strangely if I run it locally it passes. Also it seems from looking at the test that its unrelated to any of my changes. Could there be instability in the test that is causing sporadic fails ?

I found a leak in the test code. I needed to run the test suite and not just the new test to see a problem.
pushed a fix just now.

**Comment by eobrienPilz** [comment]
@iloveeclipse Checks are passing.

**Comment by eobrienPilz** [comment]
Anything else to required before it can merge ?

**Review by wimjongman — APPROVED** [review]

**Comment by wimjongman** [comment]
> Anything else to required before it can merge ?

Yes, can you comment a decent commit message?

**Comment by eobrienPilz** [comment]
> > Anything else to required before it can merge ?
> 
> Yes, can you comment a decent commit message?

I just updated the message to  "Allow customized ScopeArea implementation for FiltersConfigurationDialog"
Let me know if its not ok.

**Comment by wimjongman** [comment]
> > > Anything else to required before it can merge ?
> > 
> > 
> > Yes, can you comment a decent commit message?
> 
> I just updated the message to "Allow customized ScopeArea implementation for FiltersConfigurationDialog" Let me know if its not ok.

Please provide a brief summary of what you did to achieve "Allow customized ScopeArea implementation for FiltersConfigurationDialog"

See [Commit message reccomendations ](https://github.com/eclipse-platform/.github/blob/main/CONTRIBUTING.md#commit-message-recommendations)

**Comment by eobrienPilz** [comment]
> > > > Anything else to required before it can merge ?
> > > 
> > > 
> > > Yes, can you comment a decent commit message?
> > 
> > 
> > I just updated the message to "Allow customized ScopeArea implementation for FiltersConfigurationDialog" Let me know if its not ok.
> 
> Please provide a brief summary of what you did to achieve "Allow customized ScopeArea implementation for FiltersConfigurationDialog"
> 
> See [Commit message reccomendations ](https://github.com/eclipse-platform/.github/blob/main/CONTRIBUTING.md#commit-message-recommendations)

Updated now. Thanks for the link to comment guidelines.

**Comment by eobrienPilz** [comment]
I'm not sure why there are checks failing now. 
The only thing I changed was the commit message.

**Comment by eobrienPilz** [comment]
2 fails are AnnotationManagerTest and VirtualTableViewerTest
https://github.com/eclipse-platform/eclipse.platform.ui/pull/17/checks?check_run_id=23833989563

Does anyone have ideas why these 2 might fail now ?

Looking at results for older commits VirtualTableViewerTest at least is failing in a previous build.
https://github.com/eclipse-platform/eclipse.platform.ui/runs/23413042111

**Comment by iloveeclipse** [comment]
> Does anyone have ideas why these 2 might fail now ?

They only fail on Mac, and I believe they are not stable there. Surely unrelated to this PR.

**Comment by wimjongman** [comment]
Thanks for the patch, Enda. 🏅

**Pull request merged**
